lib/std/crypto/pbkdf2.zig

```diff
@@ -0,0 +1,280 @@
+        }
+    }
+}
+
+const htest = @import("test.zig");
+const HmacSha1 = std.crypto.auth.hmac.HmacSha1;
+
+// RFC 6070 PBKDF2 HMAC-SHA1 Test Vectors
+test "RFC 6070 one iteration" {
+    const p = "password";
+    const s = "salt";
+    const c = 1;
+    const dkLen = 20;
+
+    var derivedKey: [dkLen]u8 = undefined;
+
+    try pbkdf2(&derivedKey, p, s, c, HmacSha1);
+
+    const expected = "0c60c80f961f0e71f3a9b524af6012062fe037a6";
+
+    htest.assertEqual(expected, derivedKey[0..]);
+}
+
+test "RFC 6070 two iterations" {
+    const p = "password";
+    const s = "salt";
+    const c = 2;
+    const dkLen = 20;
+
+    var derivedKey: [dkLen]u8 = undefined;
+
+    try pbkdf2(&derivedKey, p, s, c, HmacSha1);
+
+    const expected = "ea6c014dc72d6f8ccd1ed92ace1d41f0d8de8957";
+
+    htest.assertEqual(expected, derivedKey[0..]);
+}
+
+test "RFC 6070 4096 iterations" {
+    const p = "password";
+    const s = "salt";
+    const c = 4096;
+    const dkLen = 20;
+
+    var derivedKey: [dkLen]u8 = undefined;
+
+    try pbkdf2(&derivedKey, p, s, c, HmacSha1);
+
+    const expected = "4b007901b765489abead49d926f721d065a429c1";
+
+    htest.assertEqual(expected, derivedKey[0..]);
+}
+
+test "RFC 6070 16,777,216 iterations" {
+    // These iteration tests are slow so we always skip them. Results have been verified.
+    if (true) {
+        return error.SkipZigTest;
+    }
+
+    const p = "password";
+    const s = "salt";
+    const c = 16777216;
+    const dkLen = 20;
+
+    var derivedKey = [_]u8{0} ** dkLen;
+
+    try pbkdf2(&derivedKey, p, s, c, HmacSha1);
+
+    const expected = "eefe3d61cd4da4e4e9945b3d6ba2158c2634e984";
+
+    htest.assertEqual(expected, derivedKey[0..]);
+}
+
+test "RFC 6070 multi-block salt and password" {
+    const p = "passwordPASSWORDpassword";
+    const s = "saltSALTsaltSALTsaltSALTsaltSALTsalt";
+    const c = 4096;
+    const dkLen = 25;
+
+    var derivedKey: [dkLen]u8 = undefined;
+
+    try pbkdf2(&derivedKey, p, s, c, HmacSha1);
+
+    const expected = "3d2eec4fe41c849b80c8d83662c0e44a8b291a964cf2f07038";
+
+    htest.assertEqual(expected, derivedKey[0..]);
+}
+
+test "RFC 6070 embedded NUL" {
+    const p = "pass\x00word";
+    const s = "sa\x00lt";
+    const c = 4096;
+    const dkLen = 16;
+
+    var derivedKey: [dkLen]u8 = undefined;
+
+    try pbkdf2(&derivedKey, p, s, c, HmacSha1);
+
+    const expected = "56fa6aa75548099dcc37d7f03425e0c3";
+
+    htest.assertEqual(expected, derivedKey[0..]);
+}
+
+test "Very large dkLen" {
+    // This test allocates 8GB of memory and is expected to take several hours to run.
+    if (true) {
+        return error.SkipZigTest;
+    }
+    const p = "password";
+    const s = "salt";
+    const c = 1;
+    const dkLen = 1 << 33;
+
+    var derivedKey = try std.testing.allocator.alloc(u8, dkLen);
+    defer {
+        std.testing.allocator.free(derivedKey);
+    }
+
+    try pbkdf2(derivedKey, p, s, c, HmacSha1);
+    // Just verify this doesn't crash with an overflow
```

Comment:
Would you mind verifying that the output eventually matches the expected one?

Reply:
I cannot run this test due to OOM. Im think ill try to implement mmaped-file allocator to run these kind of tests. Also i dont know what the expected output is. Its few GB long, i think we should try a checksum or something.

Reply:
If that test cannot be run, neither on CI nor on developers laptops, maybe it should be removed.

Needing a > 4GB key is completely unrealistic in any real-world scenario. So maybe the assertion in the main function could be simplified to `derivedKey.len < maxInt(u32)`. This doesn't require any testing, and ensures consistent behavior on all platforms.